File: ts/tests/conformance.test.ts
/**
 * Conformance tests loaded from docs/conformance/canonical-vectors.json.
 *
 * These tests verify that the TypeScript implementation produces identical
 * canonical forms and hashes to the Go reference implementation.
 */

import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, it, expect } from "vitest";
import {
  canonicalContentJson,
  canonicalForm,
  computeHash,
} from "../src/event.js";
import {
  Score,
  Weight,
  Activation,
  Layer,
  Cadence,
  Hash,
} from "../src/types.js";
import {
  Lifecycle,
  isValidTransition,
} from "../src/primitive.js";
import {
  ActorStatus,
  transitionTo,
} from "../src/actor.js";

// ── Load vectors ────────────────────────────────────────────────────────

const __dirname = dirname(fileURLToPath(import.meta.url));
const vectorsPath = resolve(__dirname, "..", "..", "docs", "conformance", "canonical-vectors.json");
const VECTORS = JSON.parse(readFileSync(vectorsPath, "utf-8"));

// ── Helper: build canonical form from a vector case input ───────────────

function buildCanonical(input: {
  version: number;
  id: string;
  type: string;
  source: string;
  conversation_id: string;
  timestamp_nanos: number;
  prev_hash: string;
  causes: string[];
  content: Record<string, unknown>;
}): string {
  const contentJson = canonicalContentJson(input.content);
  return canonicalForm(
    input.version,
    input.prev_hash,
    input.causes,
    input.id,
    input.type,
    input.source,
    input.conversation_id,
    input.timestamp_nanos,
    contentJson,
  );
}

// ── Canonical form tests ────────────────────────────────────────────────

describe("conformance: canonical_form (from vectors)", () => {
  const cases = VECTORS.canonical_form.cases as Array<{
    name: string;
    description: string;
    input?: Record<string, unknown>;
    expected?: Record<string, string>;
    rules?: Array<{ input: number; canonical: string }>;
    input_content?: Record<string, unknown>;
    expected_json?: string;
  }>;

  for (const tc of cases) {
    // Skip non-event cases — they are tested separately below
    if (!tc.input) continue;

    it(`${tc.name}: hash matches`, () => {
      const canon = buildCanonical(tc.input as ReturnType<typeof buildCanonical extends (a: infer T) => string ? never : never> & Parameters<typeof buildCanonical>[0]);
      const hash = computeHash(canon);

      if (tc.expected?.canonical) {
        expect(canon).toBe(tc.expected.canonical);
      }
      if (tc.expected?.hash) {
        expect(hash.value).toBe(tc.expected.hash);
      }
      if ((tc.expected as Record<string, string>)?.canonical_content_json) {
        const contentJson = canonicalContentJson((tc.input as Record<string, unknown>).content as Record<string, unknown>);
        expect(contentJson).toBe((tc.expected as Record<string, string>).canonical_content_json);
      }
    });
  }
});

// ── Number formatting ───────────────────────────────────────────────────

describe("conformance: content_json_number_formatting (from vectors)", () => {
  const numberCase = VECTORS.canonical_form.cases.find(
    (c: { name: string }) => c.name === "content_json_number_formatting",
  );

  for (const rule of numberCase.rules) {
    it(`${rule.input} => "${rule.canonical}"`, () => {
      const json = canonicalContentJson({ v: rule.input });
      expect(json).toBe(`{"v":${rule.canonical}}`);
    });
  }
});

// ── Null omission ───────────────────────────────────────────────────────

describe("conformance: content_json_null_omission (from vectors)", () => {
  const nullCase = VECTORS.canonical_form.cases.find(
    (c: { name: string }) => c.name === "content_json_null_omission",
  );

  it("null values are omitted from content JSON", () => {
    const json = canonicalContentJson(nullCase.input_content);
    expect(json).toBe(nullCase.expected_json);
  });
});

// ── Nested objects ──────────────────────────────────────────────────────

describe("conformance: content_json_nested_objects (from vectors)", () => {
  const nestedCase = VECTORS.canonical_form.cases.find(
    (c: { name: string }) => c.name === "content_json_nested_objects",
  );

  it("nested objects have sorted keys and null omission", () => {
    const json = canonicalContentJson(nestedCase.input_content);
    expect(json).toBe(nestedCase.expected_json);
  });
});

// ── Type validation ─────────────────────────────────────────────────────

describe("conformance: type_validation (from vectors)", () => {
  const typeMap: Record<string, new (v: number) => unknown> = {
    Score,
    Weight,
    Activation,
    Layer,
    Cadence,
  };

  describe("invalid values must throw", () => {
    for (const tc of VECTORS.type_validation.invalid) {
      // Hash has a string constructor, handle separately
      if (tc.type === "Hash") {
        it(`Hash(${JSON.stringify(tc.value)}) throws — ${tc.reason}`, () => {
          expect(() => new Hash(tc.value)).toThrow();
        });
        continue;
      }

      const Ctor = typeMap[tc.type as string];
      if (!Ctor) continue;

      it(`${tc.type}(${tc.value}) throws — ${tc.reason}`, () => {
        expect(() => new Ctor(tc.value)).toThrow();
      });
    }
  });

  describe("valid values must construct", () => {
    for (const tc of VECTORS.type_validation.valid) {
      const Ctor = typeMap[tc.type as string];
      if (!Ctor) continue;

      it(`${tc.type}(${tc.value}) succeeds`, () => {
        const obj = new Ctor(tc.value) as { value: number };
        expect(obj.value).toBe(tc.value);
      });
    }
  });
});

// ── Lifecycle transitions ───────────────────────────────────────────────

describe("conformance: lifecycle_transitions (from vectors)", () => {
  // Map PascalCase vector names to lowercase implementation names.
  const lifecycleMap: Record<string, string> = {
    Dormant: Lifecycle.Dormant,
    Activating: Lifecycle.Activating,
    Active: Lifecycle.Active,
    Processing: Lifecycle.Processing,
    Emitting: Lifecycle.Emitting,
    Deactivating: Lifecycle.Deactivating,
  };

  describe("LifecycleState valid transitions", () => {
    for (const [from, to] of VECTORS.lifecycle_transitions.LifecycleState.valid) {
      const fromImpl = lifecycleMap[from];
      const toImpl = lifecycleMap[to];

      it(`${from} -> ${to} is valid`, () => {
        expect(fromImpl, `unmapped lifecycle vector state: ${from}`).toBeDefined();
        expect(toImpl, `unmapped lifecycle vector state: ${to}`).toBeDefined();
        expect(isValidTransition(fromImpl, toImpl)).toBe(true);
      });
    }
  });

  describe("LifecycleState invalid transitions", () => {
    for (const [from, to] of VECTORS.lifecycle_transitions.LifecycleState.invalid) {
      const fromImpl = lifecycleMap[from];
      const toImpl = lifecycleMap[to];

      it(`${from} -> ${to} is invalid`, () => {
        expect(fromImpl, `unmapped lifecycle vector state: ${from}`).toBeDefined();
        expect(toImpl, `unmapped lifecycle vector state: ${to}`).toBeDefined();
        expect(isValidTransition(fromImpl, toImpl)).toBe(false);
      });
    }
  });

  // Map PascalCase ActorStatus vector names to enum values
  const actorStatusMap: Record<string, ActorStatus> = {
    Active: ActorStatus.Active,
    Suspended: ActorStatus.Suspended,
    Memorial: ActorStatus.Memorial,
  };

  describe("ActorStatus valid transitions", () => {
    for (const [from, to] of VECTORS.lifecycle_transitions.ActorStatus.valid) {
      const fromStatus = actorStatusMap[from];
      const toStatus = actorStatusMap[to];

      it(`${from} -> ${to} succeeds`, () => {
        expect(fromStatus, `unmapped actor status vector state: ${from}`).toBeDefined();
        expect(toStatus, `unmapped actor status vector state: ${to}`).toBeDefined();
        expect(transitionTo(fromStatus, toStatus)).toBe(toStatus);
      });
    }
  });

  describe("ActorStatus invalid transitions", () => {
    for (const [from, to] of VECTORS.lifecycle_transitions.ActorStatus.invalid) {
      const fromStatus = actorStatusMap[from];
      const toStatus = actorStatusMap[to];

      it(`${from} -> ${to} throws`, () => {
        expect(fromStatus, `unmapped actor status vector state: ${from}`).toBeDefined();
        expect(toStatus, `unmapped actor status vector state: ${to}`).toBeDefined();
        expect(() => transitionTo(fromStatus, toStatus)).toThrow();
      });
    }
  });
});

File: rust/tests/conformance_test.rs
//! Conformance tests loaded from docs/conformance/canonical-vectors.json.
//!
//! These tests verify that the Rust implementation produces identical
//! canonical forms and hashes to the Go reference implementation.

use std::collections::BTreeMap;
use std::path::PathBuf;

use serde_json::Value;

use eventgraph::event::{canonical_content_json, canonical_form, compute_hash};
use eventgraph::types::{Score, Weight, Activation, Layer, Cadence, Hash, LifecycleState};
use eventgraph::actor::ActorStatus;

// ── Load vectors ────────────────────────────────────────────────────────

fn load_vectors() -> Value {
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let vectors_path = manifest_dir
        .parent()
        .unwrap()
        .join("docs")
        .join("conformance")
        .join("canonical-vectors.json");
    let text = std::fs::read_to_string(&vectors_path)
        .unwrap_or_else(|e| panic!("Failed to read {}: {}", vectors_path.display(), e));
    serde_json::from_str(&text).expect("Failed to parse vectors JSON")
}

// ── Helper: build content BTreeMap from JSON value ──────────────────────

fn json_to_btree(obj: &Value) -> BTreeMap<String, Value> {
    let mut map = BTreeMap::new();
    if let Some(o) = obj.as_object() {
        for (k, v) in o {
            map.insert(k.clone(), v.clone());
        }
    }
    map
}

// ── Helper: build canonical from a vector case ─────────────────────────

fn build_canonical(input: &Value) -> String {
    let content = json_to_btree(&input["content"]);
    let content_json = canonical_content_json(&content);

    let causes: Vec<&str> = input["causes"]
        .as_array()
        .unwrap()
        .iter()
        .map(|v| v.as_str().unwrap())
        .collect();

    canonical_form(
        input["version"].as_u64().unwrap() as u32,
        input["prev_hash"].as_str().unwrap(),
        &causes,
        input["id"].as_str().unwrap(),
        input["type"].as_str().unwrap(),
        input["source"].as_str().unwrap(),
        input["conversation_id"].as_str().unwrap(),
        input["timestamp_nanos"].as_u64().unwrap(),
        &content_json,
    )
}

// ── Canonical form tests ────────────────────────────────────────────────

#[test]
fn conformance_vectors_bootstrap_event() {
    let vectors = load_vectors();
    let cases = vectors["canonical_form"]["cases"].as_array().unwrap();
    let tc = cases.iter().find(|c| c["name"] == "bootstrap_event").unwrap();

    let canon = build_canonical(&tc["input"]);
    let hash = compute_hash(&canon);

    assert!(canon.starts_with("1|||"));
    assert_eq!(canon, tc["expected"]["canonical"].as_str().unwrap());
    assert_eq!(hash.value(), tc["expected"]["hash"].as_str().unwrap());
}

#[test]
fn conformance_vectors_trust_updated_event() {
    let vectors = load_vectors();
    let cases = vectors["canonical_form"]["cases"].as_array().unwrap();
    let tc = cases.iter().find(|c| c["name"] == "trust_updated_event").unwrap();

    let canon = build_canonical(&tc["input"]);
    let hash = compute_hash(&canon);

    assert_eq!(hash.value(), tc["expected"]["hash"].as_str().unwrap());
}

#[test]
fn conformance_vectors_content_json_key_ordering() {
    let vectors = load_vectors();
    let cases = vectors["canonical_form"]["cases"].as_array().unwrap();
    let tc = cases.iter().find(|c| c["name"] == "content_json_key_ordering").unwrap();

    let canon = build_canonical(&tc["input"]);
    let hash = compute_hash(&canon);

    assert_eq!(hash.value(), tc["expected"]["hash"].as_str().unwrap());
}

#[test]
fn conformance_vectors_multiple_causes_sorted() {
    let vectors = load_vectors();
    let cases = vectors["canonical_form"]["cases"].as_array().unwrap();
    let tc = cases.iter().find(|c| c["name"] == "multiple_causes_sorted").unwrap();

    let canon = build_canonical(&tc["input"]);
    let hash = compute_hash(&canon);

    assert_eq!(canon, tc["expected"]["canonical"].as_str().unwrap());
    assert_eq!(hash.value(), tc["expected"]["hash"].as_str().unwrap());
}

#[test]
fn conformance_vectors_content_integer_float_formatting() {
    let vectors = load_vectors();
    let cases = vectors["canonical_form"]["cases"].as_array().unwrap();
    let tc = cases.iter().find(|c| c["name"] == "content_integer_float_formatting").unwrap();

    let content = json_to_btree(&tc["input"]["content"]);
    let content_json = canonical_content_json(&content);
    assert_eq!(content_json, tc["expected"]["canonical_content_json"].as_str().unwrap());

    let canon = build_canonical(&tc["input"]);
    let hash = compute_hash(&canon);
    assert_eq!(hash.value(), tc["expected"]["hash"].as_str().unwrap());
}

// ── Number formatting rules ─────────────────────────────────────────────

#[test]
fn conformance_vectors_number_formatting_rules() {
    let vectors = load_vectors();
    let cases = vectors["canonical_form"]["cases"].as_array().unwrap();
    let tc = cases.iter().find(|c| c["name"] == "content_json_number_formatting").unwrap();

    for rule in tc["rules"].as_array().unwrap() {
        let input = rule["input"].as_f64().unwrap();
        let expected = rule["canonical"].as_str().unwrap();

        let mut content = BTreeMap::new();
        content.insert(
            "v".to_string(),
            Value::Number(serde_json::Number::from_f64(input).unwrap()),
        );
        let json = canonical_content_json(&content);
        assert_eq!(json, format!("{{\"v\":{expected}}}"), "Failed for input {input}");
    }
}

// ── Null omission ───────────────────────────────────────────────────────

#[test]
fn conformance_vectors_null_omission() {
    let vectors = load_vectors();
    let cases = vectors["canonical_form"]["cases"].as_array().unwrap();
    let tc = cases.iter().find(|c| c["name"] == "content_json_null_omission").unwrap();

    let content = json_to_btree(&tc["input_content"]);
    let json = canonical_content_json(&content);
    assert_eq!(json, tc["expected_json"].as_str().unwrap());
}

// ── Nested objects ──────────────────────────────────────────────────────

#[test]
fn conformance_vectors_nested_objects() {
    let vectors = load_vectors();
    let cases = vectors["canonical_form"]["cases"].as_array().unwrap();
    let tc = cases.iter().find(|c| c["name"] == "content_json_nested_objects").unwrap();

    let content = json_to_btree(&tc["input_content"]);
    let json = canonical_content_json(&content);
    assert_eq!(json, tc["expected_json"].as_str().unwrap());
}

// ── Type validation ─────────────────────────────────────────────────────

#[test]
fn conformance_vectors_type_validation_invalid() {
    let vectors = load_vectors();
    let invalid = vectors["type_validation"]["invalid"].as_array().unwrap();

    for tc in invalid {
        let type_name = tc["type"].as_str().unwrap();
        let reason = tc["reason"].as_str().unwrap();

        match type_name {
            "Score" => {
                let val = tc["value"].as_f64().unwrap();
                assert!(Score::new(val).is_err(), "Score({val}) should fail: {reason}");
            }
            "Weight" => {
                let val = tc["value"].as_f64().unwrap();
                assert!(Weight::new(val).is_err(), "Weight({val}) should fail: {reason}");
            }
            "Activation" => {
                let val = tc["value"].as_f64().unwrap();
                assert!(Activation::new(val).is_err(), "Activation({val}) should fail: {reason}");
            }
            "Layer" => {
                let val = tc["value"].as_i64().unwrap();
                // Layer takes u8, but invalid values may be negative
                if val < 0 || val > 255 {
                    // Out of u8 range — would fail at conversion level
                    assert!(true);
                } else {
                    assert!(Layer::new(val as u8).is_err(), "Layer({val}) should fail: {reason}");
                }
            }
            "Cadence" => {
                let val = tc["value"].as_u64().unwrap_or(0);
                assert!(Cadence::new(val as u32).is_err(), "Cadence({val}) should fail: {reason}");
            }
            "Hash" => {
                let val = tc["value"].as_str().unwrap();
                assert!(Hash::new(val).is_err(), "Hash({val:?}) should fail: {reason}");
            }
            _ => panic!("Unknown type: {type_name}"),
        }
    }
}

#[test]
fn conformance_vectors_type_validation_valid() {
    let vectors = load_vectors();
    let valid = vectors["type_validation"]["valid"].as_array().unwrap();

    for tc in valid {
        let type_name = tc["type"].as_str().unwrap();

        match type_name {
            "Score" => {
                let val = tc["value"].as_f64().unwrap();
                let s = Score::new(val).expect(&format!("Score({val}) should succeed"));
                assert_eq!(s.value(), val);
            }
            "Weight" => {
                let val = tc["value"].as_f64().unwrap();
                let w = Weight::new(val).expect(&format!("Weight({val}) should succeed"));
                assert_eq!(w.value(), val);
            }
            "Layer" => {
                let val = tc["value"].as_u64().unwrap();
                let l = Layer::new(val as u8).expect(&format!("Layer({val}) should succeed"));
                assert_eq!(l.value(), val as u8);
            }
            "Cadence" => {
                let val = tc["value"].as_u64().unwrap();
                let c = Cadence::new(val as u32).expect(&format!("Cadence({val}) should succeed"));
                assert_eq!(c.value(), val as u32);
            }
            _ => {} // Activation not in valid vectors
        }
    }
}

// ── Lifecycle transitions ───────────────────────────────────────────────

fn parse_lifecycle_state(name: &str) -> Option<LifecycleState> {
    match name {
        "Dormant" => Some(LifecycleState::Dormant),
        "Activating" => Some(LifecycleState::Activating),
        "Active" => Some(LifecycleState::Active),
        "Processing" => Some(LifecycleState::Processing),
        "Emitting" => Some(LifecycleState::Emitting),
        "Deactivating" => Some(LifecycleState::Deactivating),
        _ => None,
    }
}

fn parse_actor_status(name: &str) -> Option<ActorStatus> {
    match name {
        "Active" => Some(ActorStatus::Active),
        "Suspended" => Some(ActorStatus::Suspended),
        "Memorial" => Some(ActorStatus::Memorial),
        _ => None,
    }
}

#[test]
fn conformance_vectors_lifecycle_valid_transitions() {
    let vectors = load_vectors();
    let transitions = vectors["lifecycle_transitions"]["LifecycleState"]["valid"]
        .as_array()
        .unwrap();

    for pair in transitions {
        let from_name = pair[0].as_str().unwrap();
        let to_name = pair[1].as_str().unwrap();

        let from = parse_lifecycle_state(from_name)
            .unwrap_or_else(|| panic!("Unmapped lifecycle vector state: {from_name}"));
        let to = parse_lifecycle_state(to_name)
            .unwrap_or_else(|| panic!("Unmapped lifecycle vector state: {to_name}"));

        assert!(
            from.can_transition_to(to),
            "Expected {from_name} -> {to_name} to be valid",
        );
    }
}

#[test]
fn conformance_vectors_lifecycle_invalid_transitions() {
    let vectors = load_vectors();
    let transitions = vectors["lifecycle_transitions"]["LifecycleState"]["invalid"]
        .as_array()
        .unwrap();

    for pair in transitions {
        let from_name = pair[0].as_str().unwrap();
        let to_name = pair[1].as_str().unwrap();

        let from = parse_lifecycle_state(from_name)
            .unwrap_or_else(|| panic!("Unmapped lifecycle vector state: {from_name}"));
        let to = parse_lifecycle_state(to_name)
            .unwrap_or_else(|| panic!("Unmapped lifecycle vector state: {to_name}"));

        assert!(
            !from.can_transition_to(to),
            "Expected {from_name} -> {to_name} to be invalid",
        );
    }
}

#[test]
fn conformance_vectors_actor_status_valid_transitions() {
    let vectors = load_vectors();
    let transitions = vectors["lifecycle_transitions"]["ActorStatus"]["valid"]
        .as_array()
        .unwrap();

    for pair in transitions {
        let from_name = pair[0].as_str().unwrap();
        let to_name = pair[1].as_str().unwrap();

        let from = parse_actor_status(from_name)
            .unwrap_or_else(|| panic!("Unmapped actor status vector state: {from_name}"));
        let to = parse_actor_status(to_name)
            .unwrap_or_else(|| panic!("Unmapped actor status vector state: {to_name}"));

        assert!(
            from.transition_to(to).is_ok(),
            "Expected ActorStatus {from_name} -> {to_name} to succeed",
        );
    }
}

#[test]
fn conformance_vectors_actor_status_invalid_transitions() {
    let vectors = load_vectors();
    let transitions = vectors["lifecycle_transitions"]["ActorStatus"]["invalid"]
        .as_array()
        .unwrap();

    for pair in transitions {
        let from_name = pair[0].as_str().unwrap();
        let to_name = pair[1].as_str().unwrap();

        let from = parse_actor_status(from_name)
            .unwrap_or_else(|| panic!("Unmapped actor status vector state: {from_name}"));
        let to = parse_actor_status(to_name)
            .unwrap_or_else(|| panic!("Unmapped actor status vector state: {to_name}"));

        assert!(
            from.transition_to(to).is_err(),
            "Expected ActorStatus {from_name} -> {to_name} to fail",
        );
    }
}
